horovod/run/elastic/registration.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2020 Uber Technologies, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from __future__ import absolute_import
+
+import logging
+import threading
+
+from collections import defaultdict
+
+READY = 'READY'
+SUCCESS = 'SUCCESS'
+FAILURE = 'FAILURE'
+
+
+class WorkerStateRegistry(object):
+    def __init__(self, driver, discovered_hosts, verbose=False):
+        self._driver = driver
+        self._discovered_hosts = discovered_hosts
+        self._lock = threading.Lock()
+        self._states = {}
+        self._workers = defaultdict(set)
+        self._barrier = None
+        self._rendezvous_id = 0
+        self._verbose = verbose
+        self._size = 0
+
+    def get(self, state):
+        return self._workers[state]
+
+    def count(self, state):
+        return len(self._workers[state])
+
+    def reset(self, size):
+        with self._lock:
+            logging.info('reset workers: {}'.format(size))
+            self._states.clear()
+            self._workers.clear()
+            self._barrier = threading.Barrier(parties=size, action=self._action)
+            self._rendezvous_id += 1
+            self._size = size
+
+    def size(self):
+        return self._size
+
+    def last_rendezvous(self):
+        return self._rendezvous_id
+
+    def record_ready(self, host, slot):
+        return self._record_state(host, slot, READY)
+
+    def record_success(self, host, slot):
+        return self._record_state(host, slot, SUCCESS)
+
+    def record_failure(self, host, slot):
+        return self._record_state(host, slot, FAILURE)
+
+    def _record_state(self, host, slot, state):
+        if self._driver.finished():
+            logging.info('driver finished, ignoring registration: {}[{}] = {}'.format(host, slot, state))
+            return self._rendezvous_id
+
+        key = (host, slot)
+        with self._lock:
+            if key in self._states:
+                logging.info('key exists, reset barrier: {}[{}] = {}'.format(host, slot, state))
+                self._barrier.reset()
```

Comment:
Strange situation, why is this not an exception and why is barrier reset sufficient?

Reply:
Good question. I've added comments to elaborate on this code.  Basically, the idea here is that after a worker records READY, it is still possible that it can fail somehow before the rendezvous barrier completes.  In that case, we want it to change its vote from READY to FAILURE.

It's not sufficient to just have the failing thread update the state to FAILURE and then return, it also needs to wait for the barrier to complete so it can decide if it needs to record its results.  This necessitates that we reset the barrier, as otherwise we will have double counted the failing worker in the barrier (once for its READY thread, and once for the FAILURE thread).

Hope that makes sense, let me know if it still needs clarification (and please take a look at updated comments).

Reply:
So the reset is only for the single host? It resets the count of that host only and not the entire barrier?

Reply:
It does reset the entire barrier, but because of the logic that handles the `BrokenBarrierException`, all the other ranks will re-enter the barrier.  Only the duplicated rank will raise an exception because its state has been overridden.

Reply:
I still don't fully understand this situation. Do other reviewers understand it with the given comments only @abditag2 @nvcastet @romerojosh ?

Reply:
Yes there seems to be a short time window where a worker could fail waiting at the barrier, then it would update its state a 2nd time with failure and we don't want 2 threads coming from same worker at the barrier.
@tgaddair Do we have a unit test case for this scenario?

Reply:
Good question, yes, we have a test for this in `test_elastic_driver.py::ElasticDriverTests::test_rank_and_size_with_host_failure`.  I'll modify the comments with @nvcastet's phrasing, as I think that's more clear.